Cython/Compiler/Parsing.py

```diff
@@ -2130,24 +2177,14 @@ def p_with_items(s, is_async=False):
             condition = p_test(s)
             s.expect(')')
 
-        if s.sy == ',':
-            s.next()
-            body = p_with_items(s)
-        else:
-            body = p_suite(s)
-        return Nodes.GILStatNode(pos, state=state, body=body, condition=condition)
+        return "gil", pos, state, condition
     else:
         manager = p_test(s)
         target = None
         if s.sy == 'IDENT' and s.systring == 'as':
             s.next()
             target = p_starred_expr(s)
-        if s.sy == ',':
-            s.next()
-            body = p_with_items(s, is_async=is_async)
-        else:
-            body = p_suite(s)
-    return Nodes.WithStatNode(pos, manager=manager, target=target, body=body, is_async=is_async)
+    return "with", pos, manager, target
```

Comment:
An alternative approach (which might/might not be better) would be to create the nodes with an empty body, then fill in the body in the `p_with_items` loop. It might be a little more readable.

Reply:
It's certainly unusual to return anything but a node from a `p_*` parser function. I wouldn't mind setting the `body` after the fact. Seems simple and reasonable here.

Reply:
It turns out that there's quite a bit of logic in the constructor of `GilStatNode` that means that you have to know what `body` is when you construct it and can't set it later.

I've therefore gone back to something closer to what I had before where the construction of the nodes is delay. Although hopefully it's a little neater now.